### DISCUSSION TITLE
Node Red terminates/exits on "ETELEGRAM: 429 Too Many Requests"

### DISCUSSION THREAD

**Discussion by mtb-github** [windkh/node-red-contrib-telegrambot#389] [Q&A]
Dear community,

this is my very first post here, so please be kind.

I am using the telegrambot in Node Red on a ioBroker installation and regularly at night (roughly same time) ioBroker shows the following log entries and Node Red completely restarts:

node-red.0 | 2024-08-21 03:13:58.082 | info | Node-RED exited with 6
-- | -- | -- | --
node-red.0 | 2024-08-21 03:13:57.516 | warn | Terminated (UNCAUGHT_EXCEPTION): Without reason
node-red.0 | 2024-08-21 03:13:57.515 | info | terminating
node-red.0 | 2024-08-21 03:13:57.002 | error | Exception-Code: ETELEGRAM: ETELEGRAM: 429 Too Many Requests: retry after 5
node-red.0 | 2024-08-21 03:13:57.002 | error | Error:  ETELEGRAM: 429 Too Many Requests: retry after 5    at  /opt/iobroker/iobroker-data/node-red/node_modules/node-telegram-bot-api/src/telegram.js:316:15     at tryCatcher  (/opt/iobroker/iobroker-data/node-red/node_modules/bluebird/js/release/util.js:16:23)     at Promise._settlePromiseFromHandler  (/opt/iobroker/iobroker-data/node-red/node_modules/bluebird/js/release/promise.js:547:31)     at Promise._settlePromise  (/opt/iobroker/iobroker-data/node-red/node_modules/bluebird/js/release/promise.js:604:18)     at Promise._settlePromise0  (/opt/iobroker/iobroker-data/node-red/node_modules/bluebird/js/release/promise.js:649:10)     at Promise._settlePromises  (/opt/iobroker/iobroker-data/node-red/node_modules/bluebird/js/release/promise.js:729:18)     at _drainQueueStep  (/opt/iobroker/iobroker-data/node-red/node_modules/bluebird/js/release/async.js:93:12)     at _drainQueue  (/opt/iobroker/iobroker-data/node-red/node_modules/bluebird/js/release/async.js:86:9)     at Async._drainQueues  (/opt/iobroker/iobroker-data/node-red/node_modules/bluebird/js/release/async.js:102:5)     at Async.drainQueues [as _onImmediate]  (/opt/iobroker/iobroker-data/node-red/node_modules/bluebird/js/release/async.js:15:14)     at process.processImmediate (node:internal/timers:476:21)
node-red.0 | 2024-08-21 03:13:57.001 | error | unhandled promise rejection: ETELEGRAM: 429 Too Many Requests: retry after 5
node-red.0 | 2024-08-21 03:13:56.988 | error | Unhandled  promise rejection. This error originated either by throwing inside of  an async function without a catch block, or by rejecting a promise which  was not handled with .catch().

As far as I know, there are no messages at all waiting to get sent out to the telegram bot at this time at all. So do you guys have any idea what might be the cause here?

Appreciate you assistance,
sincerely

Tim

**Comment by windkh** [comment]
Error 429 is solved

**Comment by windkh** [comment]
_(Drafted by Claude, Anthropic's AI coding assistant. Posted by the maintainer.)_

@mtb-github — closing-the-loop comment in case you're still on an older version:

The "Node-RED exits with `Uncaught Exception: ETELEGRAM: 429 Too Many Requests`" pattern you reported was traced to two separate underlying issues that landed in different releases:

- **[V17.2.0 / PR #435](https://github.com/windkh/node-red-contrib-telegrambot/pull/435)** added a top-level catch for unhandled Promise rejections, which closed off the most common shape where a 429 from a `sendMessage` call would bubble all the way up to `process.on('uncaughtException')` and terminate Node-RED.
- **[V17.4.0 onward](https://github.com/windkh/node-red-contrib-telegrambot/releases/tag/V17.4.0)** added auto-restart on fatal bot 'error' events, so even if a 429 or any other fatal-class error did fire the 'error' event, the bot would back off and recover instead of dying silently.
- The **sender node's 429 handler** has been honouring Telegram's `retry_after` parameter for a long time — when a sendMessage hits 429, the message stays in the per-chat queue and retries after the delay Telegram supplies. So a 429 burst slows your bot down but doesn't lose messages and doesn't crash Node-RED.

Current `latest` on npm is **V17.4.11**. If you're still on something pre-V17.2.0, upgrading via Manage Palette should resolve the crash pattern entirely. If you upgrade and still see Node-RED exit on a 429, that would be a genuinely new shape and worth a fresh issue with the log block around the crash.

Closing this discussion as resolved. Please reopen if the pattern recurs on a current version.
